clvm-derive/src/from_clvm.rs

```diff
@@ -1,66 +1,309 @@
-                let fields = &fields.named;
-                field_types = fields.iter().map(|field| field.ty.clone()).collect();
-                field_names = fields
-                    .iter()
-                    .map(|field| field.ident.clone().unwrap())
-                    .collect();
-                initializer = quote!(Self { #( #field_names, )* });
+        Data::Struct(data_struct) => {
+            if clvm_attr.untagged {
+                panic!("cannot use `untagged` on a struct");
+            }
+            let macros = repr_macros(&crate_name, clvm_attr.expect_repr());
+            let field_info = fields(&data_struct.fields);
+            impl_for_struct(&crate_name, &ast, &macros, &field_info)
+        }
+        Data::Enum(data_enum) => {
+            if !clvm_attr.untagged && clvm_attr.repr == Some(Repr::Curry) {
+                panic!("cannot use `curry` on a tagged enum, since unlike other representations, each argument is wrapped");
             }
-            Fields::Unnamed(fields) => {
-                let fields = &fields.unnamed;
-                field_types = fields.iter().map(|field| field.ty.clone()).collect();
-                field_names = fields
-                    .iter()
-                    .enumerate()
-                    .map(|(i, field)| Ident::new(&format!("field_{i}"), field.span()))
-                    .collect();
-                initializer = quote!(Self( #( #field_names, )* ));
+
+            let mut next_discriminant: Expr = parse_quote!(0);
+            let mut variants = Vec::new();
+
+            for variant in data_enum.variants.iter() {
+                let field_info = fields(&variant.fields);
+                let variant_clvm_attr = parse_clvm_attr(&variant.attrs);
+
+                if variant_clvm_attr.untagged {
+                    panic!("cannot use `untagged` on an enum variant");
+                }
+
+                let repr = variant_clvm_attr
+                    .repr
+                    .unwrap_or_else(|| clvm_attr.expect_repr());
+                if !clvm_attr.untagged && repr == Repr::Curry {
+                    panic!("cannot use `curry` on a tagged enum variant, since unlike other representations, each argument is wrapped");
+                }
+
+                let macros = repr_macros(&crate_name, repr);
+                let variant_info = VariantInfo {
+                    name: variant.ident.clone(),
+                    discriminant: variant
+                        .discriminant
+                        .as_ref()
+                        .map(|(_, discriminant)| {
+                            next_discriminant = parse_quote!(#discriminant + 1);
+                            discriminant.clone()
+                        })
+                        .unwrap_or_else(|| {
+                            let discriminant = next_discriminant.clone();
+                            next_discriminant = parse_quote!(#next_discriminant + 1);
```

Comment:
You can use constants instead of integer literals for the discriminant value, and I don't think constant values are known at the time this macro is running. This all gets optimized away at compile time though.